File: src/nodetool/models/thread.py
from datetime import datetime
from nodetool.models.base_model import DBModel, DBField, create_time_ordered_uuid
from nodetool.models.condition_builder import Field


class Thread(DBModel):
    @classmethod
    def get_table_schema(cls):
        return {"table_name": "nodetool_threads"}

    id: str = DBField(hash_key=True)
    user_id: str = DBField(default="")
    created_at: datetime = DBField(default_factory=datetime.now)
    updated_at: datetime = DBField(default_factory=datetime.now)

    @classmethod
    def find(cls, user_id: str, id: str):
        thread = cls.get(id)
        if thread and thread.user_id == user_id:
            return thread
        return None

    @classmethod
    def create(cls, user_id: str, **kwargs) -> "Thread":
        return super().create(
            id=create_time_ordered_uuid(),
            user_id=user_id,
            **kwargs,
        )

    @classmethod
    def paginate(
        cls,
        user_id: str,
        limit: int = 10,
        start_key: str | None = None,
        reverse: bool = False,
    ):
        return cls.query(
            condition=Field("user_id")
            .equals(user_id)
            .and_(Field("id").greater_than(start_key or "")),
            limit=limit,
            reverse=reverse,
        )

File: packages/models/src/thread.ts
/**
 * Thread model – conversation thread container.
 *
 * Port of Python's `nodetool.models.thread`.
 */

import { eq, desc, asc } from "drizzle-orm";
import { DBModel, createTimeOrderedUuid } from "./base-model.js";
import { getDb } from "./db.js";
import { threads } from "./schema/threads.js";

export class Thread extends DBModel {
  static override table = threads;

  declare id: string;
  declare user_id: string;
  declare title: string;
  declare created_at: string;
  declare updated_at: string;

  constructor(data: Record<string, unknown>) {
    super(data);
    const now = new Date().toISOString();
    this.id ??= createTimeOrderedUuid();
    this.title ??= "";
    this.created_at ??= now;
    this.updated_at ??= now;
  }

  override beforeSave(): void {
    this.updated_at = new Date().toISOString();
  }

  /** Find a thread by id, scoped to the user. */
  static async find(userId: string, threadId: string): Promise<Thread | null> {
    const thread = await Thread.get<Thread>(threadId);
    if (!thread) return null;
    if (thread.user_id === userId) return thread;
    return null;
  }

  /** Paginate threads for a user. */
  static async paginate(
    userId: string,
    opts: { limit?: number; startKey?: string; reverse?: boolean } = {}
  ): Promise<[Thread[], string]> {
    const { limit = 50, reverse = true } = opts;
    const db = getDb();
    const rows = await db
      .select()
      .from(threads)
      .where(eq(threads.user_id, userId))
      .orderBy(reverse ? desc(threads.updated_at) : asc(threads.updated_at))
      .limit(limit + 1)

    const items = rows.map((r: Record<string, unknown>) => new Thread(r as Record<string, unknown>));
    if (items.length <= limit) return [items, ""];
    items.pop();
    const cursor = items[items.length - 1]?.id ?? "";
    return [items, cursor];
  }
}
